---
title: "Finemapping with SuSiE and COLOC"
author: "Daria"
date: "6/1/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if (!require("dplyr")) install.packages("dplyr")
if (!require("susieR")) install.packages("susieR")
if (!require("coloc")) install.packages("coloc")
if (!require("magrittr")) install.packages("magrittr")
if (!require("locuscomparer")) install.packages("locuscomparer")
```

```{r echo = T, warning=FALSE, message=FALSE, eval=TRUE}
library(dplyr)
library(susieR)
library(coloc)
library(magrittr)
library(locuscomparer)
```

# Tutorial overview

This tutorial is aimed at running colocalization analysis with SuSiE in R.
We will need:

* summary statistics file 
* precomputed LD matrix (it can be easily done with PLINK)
* rs of SNPs from the matrix 
* eQTL data

Please, set your own paths for each file:

```{r}
gwas_path <- "/media/daria/DaryaNika/Fine_mapping/sumstats_subsets/ctsb_1000kb_window.tsv"
LD_path <- "/media/daria/DaryaNika/Fine_mapping/LD_reference/1000G_calculated_de_novo_matrix_1703/ctsb_1000kb_wo_rare.ld"
LD_rs_path <- "/media/daria/DaryaNika/Fine_mapping/LD_reference/1000G_calculated_de_novo_matrix_1703/ctsb_1000kb_wo_rare.txt"
eQTL_path <- "/media/daria/DaryaNika/Fine_mapping/QTL_windows/ctsb/ctsb_Nerve_Tibial_1e-05.txt"

gwas <- read.csv(gwas_path, sep = "\t")
LD_rs <- read.table(LD_rs_path)
LD <- read.table(LD_path, row.names = as.character(LD_rs$V1), col.names = as.character(LD_rs$V1))
eqtl <- read.csv(eQTL_path, sep = "\t")
```

Firstly let's prepare gwas data for further analysis:
```{r}
# Replace NAs with 0 and turn LD into matrix
LD <- LD %>% replace(is.na(.), 0)
LD_matrix <- as.matrix(LD)

# Check for complete cases
gwas <- gwas[complete.cases(gwas), ]

# Check for cases with beta = 0
gwas <- gwas[gwas$b != 0, ]

gwas <- gwas %>% 
  mutate(s = N_cases / (N_cases + N_controls), varbeta=se^2) %>% 
  dplyr::select(rsID, A1, A2, freq, b, se, p, bp, s, varbeta) %>%  
  dplyr::rename(SNP = rsID)

# Find MA beta and MAF
gwas$b_MA <- gwas$b
gwas[gwas$freq > 0.50, ]$b_MA <- gwas[gwas$freq > 0.50, ]$b_MA *-1

gwas$MAF <- gwas$freq
gwas[gwas$freq > 0.50, ]$MAF <- 1 - gwas[gwas$freq > 0.50, ]$MAF

gwas <- gwas[gwas$SNP %in% LD_rs$V1, ]

head(gwas)
```

Now let's prepare eqtl data:

```{r}
# Check for complete cases
eqtl <- eqtl[complete.cases(eqtl), ]

# Find MA beta and MAF
eqtl$b_MA <- eqtl$b
eqtl[eqtl$A2freq > 0.50, ]$b_MA <- eqtl[eqtl$A2freq > 0.50, ]$b_MA *-1

eqtl$MAF <- eqtl$Freq
eqtl[eqtl$A2freq > 0.50, ]$MAF <- 1 - eqtl[eqtl$A2freq > 0.50, ]$MAF

# Find Z for susie
eqtl$z <- eqtl$b_MA / eqtl$SE

eqtl <- eqtl %>% 
  dplyr::select(SNP, BP, A1, A2, Gene, b, SE, MAF, b_MA, z, p)

# Select snps from scarb2_gwas according to LD snps
eqtl <- eqtl %>% 
  dplyr::filter(SNP %in% LD_rs$V1)
```

After preparation we can join GWAS and eQTL datasets in order to find shared variants:

```{r}
# join GWAS + eQTLs
gwas_eqtl <- dplyr::inner_join(gwas, eqtl, by = "SNP", suffix = c(".gwas", ".eqtl"))
head(gwas_eqtl)
```

And select only SNPs that are associated with CTSB expression:
```{r}
# Select only CTSB
gwas_eqtl_ctsb <- gwas_eqtl[gwas_eqtl$Gene == "CTSB", ]
rs_ctsb <- as.character(gwas_eqtl_ctsb$SNP)
LD_ctsb <- LD_matrix[rs_ctsb, rs_ctsb]
```

Now we run SuSiE on GWAS part...

```{r}
# Prepare gwas data for Susie
named_beta <- extract2(gwas_eqtl_ctsb, 'b_MA.gwas') %>% set_names(gwas_eqtl_ctsb$SNP)
named_varbeta <- extract2(gwas_eqtl_ctsb, 'varbeta') %>% set_names(gwas_eqtl_ctsb$SNP)

dataset1=list(beta=named_beta, 
              varbeta=named_varbeta, 
              s =gwas_eqtl_ctsb$s, 
              type="cc", 
              snp=as.character(gwas_eqtl_ctsb$SNP), 
              LD=LD_ctsb, 
              position=gwas_eqtl_ctsb$bp)

check_dataset(dataset1)

# Run Susie
S3=runsusie(dataset1, nref=503)
summary(S3)
```

...and on eQTL part:

```{r}
# Prepare gwas data for Susie
named_pvalues <- extract2(gwas_eqtl_ctsb, 'p.eqtl') %>% set_names(gwas_eqtl_ctsb$SNP)
named_MAF <- extract2(gwas_eqtl_ctsb, 'MAF.eqtl') %>% set_names(gwas_eqtl_ctsb$SNP)

dataset2=list(pvalues=named_pvalues, 
              N=503, 
              MAF=named_MAF, 
              type="quant", 
              LD=LD_ctsb, 
              z=gwas_eqtl_ctsb$z, 
              snp=as.character(gwas_eqtl_ctsb$SNP), 
              position=gwas_eqtl_ctsb$bp)

check_dataset(dataset2)

# run susie
S4=runsusie(dataset2, nref=503)
summary(S4)
```

Finally we perform colocalization analysis:

```{r}
susie.res=coloc.susie(S3,S4)
susie.res$summary[susie.res$summary$PP.H4.abf > 0.8, ]
```

We can see that posterior probability of having shared variant is higher than 80%.

```{r warning=FALSE, message=FALSE,}
# prepare datasets for locuscompare
df_sub_eqtl <- gwas_eqtl_ctsb %>% 
      dplyr::select(SNP, p.eqtl) %>%
      dplyr::mutate(SNP = as.character(SNP)) %>% 
      dplyr::rename(rsid = SNP, pval = p.eqtl)
df_sub_gwas <- gwas_eqtl_ctsb %>% 
      dplyr::select(SNP, p.gwas) %>%
      dplyr::mutate(SNP = as.character(SNP)) %>%
      dplyr::rename(rsid = SNP, pval = p.gwas)
    
    # plot significant colocalization
plot <- locuscompare(in_fn1=df_sub_gwas,
                     in_fn2=df_sub_eqtl,
                     title1="GWAS", 
                     title2="eQTL", 
                     genome="hg19", 
                     population = "EUR",
                     snp = c("rs1692821", "rs8005", "rs1296028", "rs1736065")) 

plot
```

